### PR TITLE
kubernetes: Set up basic iterated adversarial training

### PR DESCRIPTION
**Pull request by tomtseng**
Setting up basic implementation of iterated adversarial training.
Associated `KataGo-custom` PR: https://github.com/AlignmentResearch/KataGo-custom/pull/83

Details:
  * Add `--iterated-training` flag to `./launch-job.sh` to perform iterated adversarial training.
  * In each iteration, a regular training run executes until the curriculum finishes.
  * In the next iteration, the latest adversary switches with the victim. The curriculum, victimplay, and evaluate-loop configs swap to the configs in `configs/iterated-training/alternate-*` since we may need to modify the `searchAlgorithm` or number of visits. The following iterations will switch back to the original configs, then back to the alternate configs, and so on.
  * The new scripts in `kubernetes/iterated-training` are wrapper scripts around the ordinary training scripts in `kubernetes/`. The wrapper scripts check when to move to the next iteration and handle swapping configs.

Example command to run adversarial training with the first iteration fine-tuning cp505 against the cyclic-adversary:
```bash
./launch-job.sh -g 1 -m 8 \ 
  --iterated-training --alternate-iteration-first \
  --curriculum /go_attack/configs/iterated-training/curriculum.json \
  --warmstart-ckpt /shared/victim-weights/kata1-b40c256-s11840935168-d2898845681 \
  --victim-ckpt /shared/victimplay/ttseng-avoid-pass-alive-coldstart-39-20221025-175949/models/t0-s545065216-d136760487 \
  ttseng-iterated
```

### CODE REVIEW

File: configs/iterated-training/alternate-experiment.cfg [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Victimplay config used on every other iteration of iterated training.
+@include ../active-experiment.cfg
+
+searchAlgorithm0 = AMCTS
```
Comment (ed1d1a8d) [inline configs/iterated-training/alternate-experiment.cfg:4]: Maybe add a comment here denoting which one is the cyclic policy and which one is KataGo?

File: configs/iterated-training/alternate-match-1gpu.cfg [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# Evaluation config used on every other iteration of iterated training.
+@include ../match-1gpu.cfg
+
+searchAlgorithm0 = AMCTS
```
Comment (ed1d1a8d) [inline configs/iterated-training/alternate-match-1gpu.cfg:4]: Like above, maybe add a comment here denoting which one is the cyclic policy and which one is KataGo?

File: configs/iterated-training/alternate-experiment.cfg [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+searchAlgorithm1 = MCTS
```
Comment (ed1d1a8d) [inline configs/iterated-training/alternate-experiment.cfg:5]: Maybe add a comment here denoting which one is the cyclic policy and which one is KataGo, and which one is being trained and which is frozen?

File: configs/iterated-training/alternate-match-1gpu.cfg [hunk 4]
```diff
@@ -0,0 +1,5 @@
...
+searchAlgorithm1 = MCTS
```
Comment (ed1d1a8d) [inline configs/iterated-training/alternate-match-1gpu.cfg:5]: Like above, maybe add a comment here denoting which one is the cyclic policy and which one is KataGo, and which one is being trained and which is frozen?

File: kubernetes/launch-job.sh [hunk 5]
```diff
@@ -6,14 +6,18 @@
...
+DEFAULT_ALTERNATE_CURRICULUM="/go_attack/configs/iterated-training/alternate-curriculum.json"
 DEFAULT_LR_SCALE=1.0
 DEFAULT_NUM_VICTIMPLAY_GPUS=4
 
 usage() {
   echo "Usage: $0 [--victimplay-gpus GPUS] [--victimplay-max-gpus MAX_GPUS]"
-  echo "          [--curriculum CURRICULUM] [--gating] [--lr-scale]"
-  echo "          [--predictor] [--predictor-warmstart-ckpt CHECKPOINT]"
-  echo "          [--resume TIMESTAMP] [--warmstart-ckpt CHECKPOINT]"
+  echo "          [--iterated-training] [alternate-iteration-first]"
```
Comment (ed1d1a8d) [inline kubernetes/launch-job.sh:15]:
```suggestion
  echo "          [--iterated-training] [--alternate-iteration-first]"
```

File: kubernetes/launch-job.sh [hunk 6]
```diff
@@ -53,14 +66,29 @@ usage() {
   echo "    /shared/victimplay/ttseng-avoid-pass-alive-coldstart-39-20221025-175949/models/t0-s545065216-d136760487"
   echo "    for the cyclic-adversary. The corresponding model file is expected"
   echo "    to be in either the TF weights directory or /shared/victims/."
+  echo "  --victim-ckpt CHECKPOINT"
+  echo "    Only for and is required for iterated training. Path to the TF weights"
+  echo "    directory for the last victim in the initial curriculum."
```
Comment (ed1d1a8d) [inline kubernetes/launch-job.sh:71]: Am a little confused about this naming due to the victim actually alternating between the cyclic policy and KataGo. Which victim is this referring to?

**Review by ed1d1a8d — CHANGES_REQUESTED** [review]
This was a pretty complicated set of scripts 😨 
Looks good for now though we should definitely figure out what is going on with train-eval discrepancies as well as the cyclic-policy being unable to be finetuned on iteration 1.

Wondering too if the launch scripts could be simplified if we converted them to python...

File: kubernetes/launch-job.sh [hunk 7]
```diff
@@ -67,8 +67,10 @@ usage() {
   echo "    for the cyclic-adversary. The corresponding model file is expected"
   echo "    to be in either the TF weights directory or /shared/victims/."
   echo "  --victim-ckpt CHECKPOINT"
-  echo "    Only for and is required for iterated training. Path to the TF weights"
-  echo "    directory for the last victim in the initial curriculum."
+  echo "    Only for and is required for iterated training. Path to the TF"
+  echo "    weights directory for the last victim (i.e., the bot not"
+  echo "    being trained) in the initial iteration's curriculum (CURRICULUM",
+  echo "    or ALTERNATE_CURRICULUM if --alternate-iteration-first is set)."
```
Comment (ed1d1a8d) [inline kubernetes/launch-job.sh:73]: Does the name of this checkpoint need to match the name given in the curriculum file?

**Review by ed1d1a8d — CHANGES_REQUESTED** [review]
Just had one additional question.

**Review by tomtseng — COMMENTED** [review]
(oops, I had written these comments when I opened the PR but forgot to post them)

File: kubernetes/launch-job.sh (hunk 7 above)
Comment (tomtseng) [reply kubernetes/launch-job.sh]: just needs to be a directory containing the correct TF weights for the model, I don't think the names have to match (e.g., things getting renamed should be fine)

**Review by ed1d1a8d — APPROVED** [review]
LGTM. Thanks for putting these scripts together!

**Pull request merged**